### PR TITLE
Settings system (without TMPro package updates)

### PR DESCRIPTION
**Pull request by Acimaz**
This time we have a clean settings system PR without TMPro stuff in the commits.
Beside the fact that my UI eventsystem is still using the default input action map provided by the Input package here are the listed changes:

- Implemented General tab - Language setting

- Implemented Graphics tab - Graphics Presets, Resolution (all supported resolutions on users device), fullscreen mode settings

- Implemented Advanced Graphics - Anisotropic filtering, Anti-Aliasing, Shadow Distance, Shadow Quality settings

- Implemented Audio tab - Music Volume, Sfx Volume settings

- Moved SettingsSystem scripts to Scripts/Systems/Settings (if we use asmdef files in the future this hiararchy might be better than throwin every script under the scripts folder)

- Added scriptable object which holds graphics presets

- Updated SettingsSystem script to utilize the presets object

- Logic is now split into sub components: Genera, Graphics, Audio and Input (currently empty as input system is wip)

- The settings system class can be used to get information from all sub components, a distributor

- Fixed an issue with Screen resolution not being correct when in windowed mode (Screen.currentResolution gives back desktop resolution when in windowed mode)

- Added Input UI elements which are placeholders for now, once the input system is done this can be populated like the other settings sub components

The same and further info is in the commits.

### CODE REVIEW

**Comment by Acimaz** [comment]
As you mentioned this https://github.com/UnityTechnologies/open-project-1/pull/30#issuecomment-703091799 @ciro-unity 

Do you want me to add the necessary bindings to the "GameInput" Input Action Map in the project so it works with the UI? Its nothing difficult but I don't want to violate any restriction or WIP by someone else in the unity staff.

**Comment by konsnos** [comment]
> without TMPro stuff in the commits

I haven't followed the previous conversation but judging from this, it seems that `SettingsSystemGeneralComponent.cs` and `SettingsSystemGraphicsComponent.cs` still have TMPro stuff in it.

**Comment by Acimaz** [comment]
> > without TMPro stuff in the commits
> 
> I haven't followed the previous conversation but judging from this, it seems that `SettingsSystemGeneralComponent.cs` and `SettingsSystemGraphicsComponent.cs` still have TMPro stuff in it.

Yes of course, those are references to textmesh pro objects in the scene, which are needed to display information in the UI. The previous conversation was targeting the import of TMPro essentials and the update of TMPro via the package manager, which currently is merged from another PR into the master branch. For more info please read the original settings system PR comments (#30).

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsPresetsScriptableObject.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        public float shadowDistance;
+        public bool custom;
+    }
+
+    public enum GraphicsQualityLevel
+    {
+        Low,
+        Middle,
+        High
+    }
```
Comment (cooljmx) [inline UOP1_Project/Assets/Scripts/Systems/Settings/SettingsPresetsScriptableObject.cs:27]: Why these contracts are nested?

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        MusicVolume = 50f;
+        musicVolumeSlider.SetValueWithoutNotify(MusicVolume);
+        //TODO: load previous sfx volume setting
+        SfxVolume = 50f;
+        sfxVolumeSlider.SetValueWithoutNotify(SfxVolume);
+    }
+    
+    #region UI CALLBACKS
+    
+    //TODO: clamp volume to [0, 1] or [0, 100]? Change Slider min max value in editor depending on use case
```
Comment (cooljmx) [inline UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs:31]: Comments in code is a bad idea. IMHO it's better to create a ticket or a new pull request.

Comment (davejrodriguez) [reply UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs:31]: [Open Projects - Conventions](https://docs.google.com/document/d/1-eUWZ0lWREFu5iH-ggofwnixDDQqalOoT4Yc0NpWR3k/edit)

Comments are fine as long as they're not redundent. And TODOs in particular are allowed as long as the functionality is not broken.

The real issue is the use of `#region`.

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemGeneralComponent.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    [SerializeField] TMP_Dropdown languageDropdown;
+
+    SettingsSystem settingsSystem;
+    public LanguageSetting Language { get; private set; }
+    
+    public enum LanguageSetting
+    {
+        English,
+        German,
+        //TODO: which languages are going to be supported?
```
Comment (cooljmx) [inline UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemGeneralComponent.cs:18]: Why did you ask this question in code? Please create a new thread here https://forum.unity.com/forums/open-projects.531/

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs (hunk 2 above)
Comment (cooljmx) [reply UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs:31]: Comments make code unreadable. No one will support comments actuality

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemGeneralComponent.cs (hunk 3 above)
Comment (Acimaz) [reply UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemGeneralComponent.cs:18]: This was a reminder for myself actually. And as its not that an important question I didn't ask it in the forums for now.

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs (hunk 2 above)
Comment (Acimaz) [reply UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs:31]: I think comments don't make the code unreadable. You were able to read and understand it obviously. Everyone uses comments and so do I. I even think they can contribute to make code more understandable in certain circumstances.

Also those conventions linked above were made after I created the PR and I didn't know that using regions, which unity itself and a lot of other companies make use of, is forbidden.

If so I will remove them, doesn't harm in any way either way.

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsPresetsScriptableObject.cs (hunk 1 above)
Comment (Acimaz) [reply UOP1_Project/Assets/Scripts/Systems/Settings/SettingsPresetsScriptableObject.cs:27]: What exactly do you mean by contracts?

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs (hunk 2 above)
Comment (cooljmx) [reply UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs:31]: Just read Clean code by uncle Bob please

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemGeneralComponent.cs (hunk 3 above)
Comment (cooljmx) [reply UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemGeneralComponent.cs:18]: Imho language support is very important.

Comment (davejrodriguez) [reply UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemGeneralComponent.cs:18]: It's unofficially planned: https://forum.unity.com/threads/localization-system.982776/

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs (hunk 2 above)
Comment (Acimaz) [reply UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemAudioComponent.cs:31]: I see. Your only issue is that I don't follow some code conventions of an individual (uncle bob), of which you think is "the only way". Thank you for your constructive comment, completely ignoring what I said. 👍

**Comment by MEGATANK58** [comment]
I would suggest closing this PR as it's not necessary or being used now

**Comment by Acimaz** [comment]
> I would suggest closing this PR as it's not necessary or being used now

A lot of PRs don't get a response at the moment. Until then, I'd say not to close it and wait for a response from the unity staff.

File: UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemGraphicsComponent.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        antiAliasingText.text = currentAdvancedGraphics.antiAliasing.ToString();
+        fullscreenToggle.SetIsOnWithoutNotify(FullScreen);
+        graphicsPresetsDropdown.SetValueWithoutNotify(currentAdvancedGraphics.custom ? graphicsPresetsDropdown.options.Count-1 : currentQualityIndex);
+    }
+
+    #region UI CALLBACKS
+
+    public void OnChangeFullscreen(bool fullScreen)
+    {
+        Screen.fullScreenMode = fullScreen ? FullScreenMode.ExclusiveFullScreen : FullScreenMode.Windowed;
```
Comment (f-montanari) [inline UOP1_Project/Assets/Scripts/Systems/Settings/SettingsSystemGraphicsComponent.cs:101]: Unity seems to support Borderless Windowed mode aswell (see FullScreenWindow in [FullScreenMode](https://docs.unity3d.com/ScriptReference/FullScreenMode.html)). I don't know if this option is already defined to be just a toggle, but we might want to support this option later.

**Comment by ciro-unity** [comment]
I was reviewing this PR and it has a lot of good stuff, I wanted to merge it, but... I don't want to merge it just yet because it integrates UI, and we A) haven't done wireframing yet and B) we're not sure yet how are we going to implement selection, and this UI is really thought for the mouse only.

So I'll keep it open, probably take some code from it later on.

**Comment by uChema** [comment]
Hello Here ! I'm reviewing this PR and it has a lot of good stuff! 
Now that we have the UI and can start implementing the solution I'm going to use part of your code, given that some of the initial requirement was changed (eg-we don't have input rebinding anymore).

**Pull request merged**